### PR TITLE
f32/f64: add ConvolveDecimate fused strided convolution

### LINKED ISSUE #51: f32/f64: add decimating (strided) valid convolution for FIR downsampling
## Summary

Add a decimating valid-convolution primitive to `f32` and `f64`:

```go
// ConvolveDecimate keeps only every factor-th valid-convolution output,
// starting at phase:
//   dst[k] = sum_{i=0}^{len(kernel)-1} signal[phase + k*factor + i] * kernel[i]
func ConvolveDecimate(dst, signal, kernel []float32, factor, phase int)
```

(and the `f64` twin).

## Motivation

go-audio-resampler's integer-ratio downsampler (`DFTDecimationStage.processZeroCopy`, `internal/engine/dft_stage.go`) runs:

```go
for pos := decimPhase; pos < numFilterable; pos += factor {
    out[k] = ops.DotProductUnsafe(history[pos:pos+numTaps], coeffs) // ~241 taps
}
```

That is one indirect SIMD call per kept output, and the coefficient vector is reloaded on every call. A decimating-convolution primitive fuses the whole loop into one call and keeps the kernel resident across outputs.

This is the 96 kHz -> 48 kHz (2:1) and 192 kHz -> 48 kHz (4:1) path. birdnet-go downsamples native high-rate sources to 48 kHz on the float32 streaming path, so this is a real hot loop.

## Semantics

- Kernel is applied as a plain dot product (the caller pre-reverses for convolution, as the resampler already does). Keeping it a plain dot matches the existing `DotProductUnsafe` usage.
- `factor >= 1` (factor == 1 is valid convolution at every position). `phase` in `[0, factor)`.
- Output count = number of strided positions `k` with `phase + k*factor + len(kernel) - 1 < len(signal)`; the primitive writes `min(len(dst), thatCount)`.

## Implementation notes

- Reuse the existing dotProduct kernels; push the outer stride loop into asm so the kernel stays in registers across outputs (the main win over the current Go-level loop of indirect calls).
- All ISA backends (AVX-512 / AVX+FMA / SSE, NEON) plus a pure-Go fallback.
- Allocation-free; caller owns `dst`.

## Alternative already in the library

`DotProductBatch(results, rows, vec)` (`f32/f32.go:173`) could serve this if the caller builds a `[][]F` of window sub-slices, but the sliding-window stride case is cleaner and avoids the per-call slice-of-slices as a dedicated primitive.

## Tests and acceptance

- Parity vs scalar reference (a loop of dot products) for factor in {1,2,3,4}, phase in `[0,factor)`, tap counts {16,20,32,64,128,241}, assorted signal lengths and unaligned tails.
- `asmcheck_test.go` entries for new asm; benchmark vs the current per-output `DotProductUnsafe` loop.
- Allocation-free.

Scoped to a single PR. Planning only; implementation in a separate session.

### PR DESCRIPTION
**Pull request by tphakala**
## What

Adds `ConvolveDecimate(dst, signal, kernel []T, factor, phase int)` to `f32` and `f64`, a decimating (strided) valid convolution that keeps every `factor`-th valid-convolution output starting at `phase`:

```
dst[k] = sum_{i} signal[phase + k*factor + i] * kernel[i]
```

With `factor == 1, phase == 0` it is exactly `ConvolveValid`. This implements the first PR-sized item from the resampling tracking issue (#54): the integer-ratio FIR downsample path (96k/192k -> 48k) that go-audio-resampler runs today as a Go loop of per-window `DotProductUnsafe` calls. The fused primitive collapses that loop into one allocation-free call.

## How

The outer stride loop lives in assembly, holding the kernel pointer/length and the signal offset in registers across outputs. The inner dot product replicates the existing `dotProduct` kernels exactly, so every output is **bit-identical** to a per-window `DotProductUnsafe`.

- Backends: **AVX-512, AVX+FMA, SSE2** (amd64) and **NEON** (arm64), plus a pure-Go fallback.
- Allocation-free; operates on caller-provided buffers (preserves the zero-alloc streaming contract).
- The arm64 dispatch mirrors `dotProduct`'s NEON length threshold (>= 4 for f32, >= 2 for f64) so the fused kernel and a per-window dot select the same backend, keeping results identical.

## Performance

Speedup over the per-window `DotProductUnsafe` loop (the realistic consumer baseline; signal length 4096, allocation-free). The win is largest for short kernels, where the baseline's per-output call, dispatch and `VZEROUPPER` overhead dominates.

| Config | f32 x86 (AVX2) | f64 x86 (AVX2) | f32 NEON (Pi 5) | f64 NEON (Pi 5) |
|--------|------|------|------|------|
| 20 taps, 2x | 2.0x | 2.3x | 1.7x | 1.9x |
| 32 taps, 2x | 2.3x | 1.7x | 1.9x | 1.7x |
| 64 taps, 2x | 2.0x | 1.8x | 1.7x | 1.3x |
| 241 taps, 2x | 1.5x | 1.3x | 1.2x | 1.1x |
| 241 taps, 4x | 1.4x | 1.2x | 1.2x | 1.1x |

## Validation

- **Bit-exact oracle**: output is bit-identical to a `DotProductUnsafe` loop at the same strided windows, checked across factors {1..4}, every phase, tap counts {1,4,7,16,20,32,64,128,241} and ragged signal lengths.
- **Per-backend**: a pointer-swap test exercises every kernel the host CPU can run: Go/SSE/AVX (f32) and Go/SSE2/AVXNoFMA/AVX (f64) on this x86-64 host. The **NEON f32 and f64 kernels were run and validated bit-exact on real aarch64 hardware** (Raspberry Pi 5).
- **AVX-512** is build- and `go vet` (asmdecl) checked; it will run on capable CI.
- Scalar-reference parity, edge cases (empty kernel, `factor < 1`, negative/past-end phase, short dst), allocation-free, and `-race` all pass.

Closes #51. Part of #54.

## Summary by CodeRabbit

* **New Features**
  * Added ConvolveDecimate: decimating (strided) valid convolution API with optimized implementations and fallbacks.

* **Documentation**
  * Added docs and a performance subsection with benchmarks comparing SIMD backends across kernel sizes and decimation factors.

* **Tests**
  * Added extensive cross-ISA tests and benchmarks validating correctness, edge cases, bit-identical backends, allocation behavior, and performance.

### CODE REVIEW

File: f32/f32_amd64.s [hunk 1]
```diff
@@ -4439,3 +4439,275 @@ addsub_scalar:
...
+// reduction) so results are bit-identical to a per-window DotProductUnsafe.
+// Outer state lives in R8-R14; the inner dot uses SI/DI/CX/AX and Y0-Y5.
+TEXT ·convolveDecimateAVX(SB), NOSPLIT, $0-88
+    MOVQ dst_base+0(FP), R8        // output pointer
+    MOVQ dst_len+8(FP), R9         // n outputs
+    MOVQ signal_base+24(FP), R10   // signal base
+    MOVQ kernel_base+48(FP), R11   // kernel base
+    MOVQ kernel_len+56(FP), R12    // kLen
+    MOVQ factor+72(FP), R13        // factor (elements)
+    MOVQ phase+80(FP), R14         // pos (elements)
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f32/f32_amd64.s]: Switched pos from R14 to BX here in 1ca56a6. R14 is the g register, agreed. For accuracy though: I verified these specific leaf NOSPLIT kernels do not actually crash today. The ABIInternal->ABI0 call wrapper restores g into R14 from TLS on return, and stackmap-less hand-written asm is never async-preempted, so the preemption handler never reads the clobbered R14 mid-loop. I confirmed both empirically on go1.26 (a 2M-iteration alloc loop and a 5000x2M long-spin loop under forced GC, neither crashed). So the severity is latent, not an active crash. BX is still the right fix: it follows the documented amd64 ABI convention, removes the footgun the moment a future edit adds a CALL into the runtime or a stack map, and is one byte shorter (no REX prefix). Also documented the convention in CLAUDE.md.

File: f32/f32_amd64.s [hunk 2]
```diff
@@ -4439,3 +4439,275 @@ addsub_scalar:
...
+// dotProductAVX512 (4 Z accumulators, 64/16/scalar reduction) for bit-identical
+// results vs a per-window DotProductUnsafe on the AVX-512 path.
+TEXT ·convolveDecimateAVX512(SB), NOSPLIT, $0-88
+    MOVQ dst_base+0(FP), R8
+    MOVQ dst_len+8(FP), R9
+    MOVQ signal_base+24(FP), R10
+    MOVQ kernel_base+48(FP), R11
+    MOVQ kernel_len+56(FP), R12
+    MOVQ factor+72(FP), R13
+    MOVQ phase+80(FP), R14
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f32/f32_amd64.s]: Fixed in 1ca56a6: pos moved from R14 to BX in this function too.

File: f32/f32_amd64.s [hunk 3]
```diff
@@ -4439,3 +4439,275 @@ addsub_scalar:
...
+// (single accumulator, 4-wide loop, SHUFPS horizontal sum) for bit-identical
+// results vs a per-window DotProductUnsafe on the SSE path.
+TEXT ·convolveDecimateSSE(SB), NOSPLIT, $0-88
+    MOVQ dst_base+0(FP), R8
+    MOVQ dst_len+8(FP), R9
+    MOVQ signal_base+24(FP), R10
+    MOVQ kernel_base+48(FP), R11
+    MOVQ kernel_len+56(FP), R12
+    MOVQ factor+72(FP), R13
+    MOVQ phase+80(FP), R14
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f32/f32_amd64.s]: Fixed in 1ca56a6: pos moved from R14 to BX in this function too.

File: f64/f64_amd64.s [hunk 4]
```diff
@@ -3921,3 +3921,265 @@ round_avx_scalar:
...
+// reduction) so results are bit-identical to a per-window DotProductUnsafe.
+// Outer state lives in R8-R14; the inner dot uses SI/DI/CX/AX and Y0-Y5.
+TEXT ·convolveDecimateAVX(SB), NOSPLIT, $0-88
+    MOVQ dst_base+0(FP), R8        // output pointer
+    MOVQ dst_len+8(FP), R9         // n outputs
+    MOVQ signal_base+24(FP), R10   // signal base
+    MOVQ kernel_base+48(FP), R11   // kernel base
+    MOVQ kernel_len+56(FP), R12    // kLen
+    MOVQ factor+72(FP), R13        // factor (elements)
+    MOVQ phase+80(FP), R14         // pos (elements)
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f64/f64_amd64.s]: Fixed in 1ca56a6: pos moved from R14 to BX in this function too.

File: f64/f64_amd64.s [hunk 5]
```diff
@@ -3921,3 +3921,265 @@ round_avx_scalar:
...
+// dotProductAVX512 (4 Z accumulators, 32/8/scalar reduction) for bit-identical
+// results vs a per-window DotProductUnsafe on the AVX-512 path.
+TEXT ·convolveDecimateAVX512(SB), NOSPLIT, $0-88
+    MOVQ dst_base+0(FP), R8
+    MOVQ dst_len+8(FP), R9
+    MOVQ signal_base+24(FP), R10
+    MOVQ kernel_base+48(FP), R11
+    MOVQ kernel_len+56(FP), R12
+    MOVQ factor+72(FP), R13
+    MOVQ phase+80(FP), R14
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f64/f64_amd64.s]: Fixed in 1ca56a6: pos moved from R14 to BX in this function too.

File: f64/f64_amd64.s [hunk 6]
```diff
@@ -3921,3 +3921,265 @@ round_avx_scalar:
...
+// (single accumulator, 2-wide loop, SHUFPD horizontal sum) for bit-identical
+// results vs a per-window DotProductUnsafe on those paths.
+TEXT ·convolveDecimateSSE2(SB), NOSPLIT, $0-88
+    MOVQ dst_base+0(FP), R8
+    MOVQ dst_len+8(FP), R9
+    MOVQ signal_base+24(FP), R10
+    MOVQ kernel_base+48(FP), R11
+    MOVQ kernel_len+56(FP), R12
+    MOVQ factor+72(FP), R13
+    MOVQ phase+80(FP), R14
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f64/f64_amd64.s]: Fixed in 1ca56a6: pos moved from R14 to BX in this function too.

File: f32/convolve_decimate_test.go [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+		t.Errorf("phase past end wrote to dst: %v", dst)
+	}
+
+	// dst shorter than available outputs: only len(dst) written.
+	dst = []float32{0}
+	ConvolveDecimate(dst, signal, []float32{1}, 1, 0) // 5 outputs available
+	if dst[0] != 1 {
+		t.Errorf("dst[0] = %v, want 1", dst[0])
+	}
+}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f32/convolve_decimate_test.go:151]: Added the phase>=factor no-op case in 1ca56a6 (also an empty-dst n==0 case).

File: f32/f32.go [hunk 8]
```diff
@@ -195,6 +195,37 @@ func ConvolveValid(dst, signal, kernel []float32) {
...
+// and phase == 0 this is exactly ConvolveValid.
+//
+// The number of outputs is the count of strided positions whose full kernel
+// window fits in signal; ConvolveDecimate writes min(len(dst), that count) and
+// leaves the remainder of dst untouched. It allocates nothing and operates on
+// the caller-provided buffers.
+func ConvolveDecimate(dst, signal, kernel []float32, factor, phase int) {
+	kLen := len(kernel)
+	if kLen == 0 || factor < 1 || phase < 0 {
+		return
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f32/f32.go:216]: Added `|| phase >= factor` to the guard in 1ca56a6 and updated the doc comment's no-op list.

File: f64/convolve_decimate_test.go [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+	if dst[0] != 99 {
+		t.Errorf("phase past end wrote to dst: %v", dst)
+	}
+
+	dst = []float64{0}
+	ConvolveDecimate(dst, signal, []float64{1}, 1, 0)
+	if dst[0] != 1 {
+		t.Errorf("dst[0] = %v, want 1", dst[0])
+	}
+}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f64/convolve_decimate_test.go:132]: Added the phase>=factor no-op case in 1ca56a6 (also an empty-dst n==0 case).

File: f64/convolve_decimate_test.go [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+	rng := rand.New(rand.NewSource(0x5117D))
+	randSlice := func(n int) []float64 {
+		s := make([]float64, n)
+		for i := range s {
+			s[i] = rng.Float64()*2 - 1
+		}
+		return s
+	}
+
+	taps := []int{1, 3, 5, 8, 16, 20, 32, 64, 128, 241}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f64/convolve_decimate_test.go]: Updated both f64 tap grids to {1,4,7,16,20,32,64,128,241} in 1ca56a6 so the 4- and 7-tap cases are exercised.

File: f64/f64.go [hunk 11]
```diff
@@ -339,6 +339,37 @@ func ConvolveValid(dst, signal, kernel []float64) {
...
+// and phase == 0 this is exactly ConvolveValid.
+//
+// The number of outputs is the count of strided positions whose full kernel
+// window fits in signal; ConvolveDecimate writes min(len(dst), that count) and
+// leaves the remainder of dst untouched. It allocates nothing and operates on
+// the caller-provided buffers.
+func ConvolveDecimate(dst, signal, kernel []float64, factor, phase int) {
+	kLen := len(kernel)
+	if kLen == 0 || factor < 1 || phase < 0 {
+		return
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f64/f64.go:360]: Added `|| phase >= factor` to the guard in 1ca56a6 and updated the doc comment's no-op list.

File: f32/convolve_decimate_amd64_test.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+	backends := []struct {
+		name string
+		init func()
+		ok   bool
+	}{
+		{"Go", initGo, true},
+		{"SSE", initSSE, cpu.X86.SSE2},
+		{"AVX", initAVX, cpu.X86.AVX && cpu.X86.FMA},
+		{"AVX512", initAVX512, cpu.X86.AVX512F && cpu.X86.AVX512VL},
+	}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f32/convolve_decimate_amd64_test.go:36]: Not applicable to f32. The f32 AVX kernels use VFMADD231PS, so f32 has no AVX-without-FMA path: init() in f32_amd64.go selects initAVX only when cpu.X86.AVX && cpu.X86.FMA, and AVX-but-no-FMA CPUs fall through to initSSE. There is no initAVXNoFMA in the f32 package, so the suggested table entry would not compile. AVX-without-FMA hosts are therefore already covered by the existing SSE backend entry. (f64 does have a real AVX-no-FMA path, and its backend table already includes the AVXNoFMA entry.)

File: f32/f32_amd64.s [hunk 13]
```diff
@@ -4439,3 +4439,277 @@ addsub_scalar:
...
+//
+// Fused decimating valid convolution: for each output k it computes the dot
+// product of signal[pos:pos+kLen] with kernel, then advances pos by factor.
+// The inner dot replicates dotProductAVX exactly (4 accumulators, 32/8/scalar
+// reduction) so results are bit-identical to a per-window DotProductUnsafe.
+// Outer state lives in R8-R13 plus BX (pos); the inner dot uses SI/DI/CX/AX and
+// Y0-Y5. BX (not R14) holds pos: R14 is the goroutine g pointer in Go's amd64
+// ABI and must not be clobbered.
+TEXT ·convolveDecimateAVX(SB), NOSPLIT, $0-88
+    MOVQ dst_base+0(FP), R8        // output pointer
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f32/f32_amd64.s:4453]: Keeping R8 here. This differs from the R14 case: R14 is reserved for the goroutine g pointer, so that was a correctness/convention fix. R8 is a normal scratch register in Go's amd64 ABI, so there is no ABI issue. The swap to DX is purely instruction size: it saves about 3 bytes per function (the load, store and pointer advance) with zero runtime effect, since these kernels are not instruction-fetch bound and the hot path is the inner dot over SI/DI and the Y registers. I measured the analogous R14->BX swap with benchstat (8 runs each side) and it was statistical noise (p=0.31-0.94, geomean +0.24%), so this would be too. The function comment documents 'outer state lives in R8-R13' as a deliberate, readable grouping; swapping only R8 to DX while R9-R13 stay extended trades that clarity for an immeasurable size win, so I'd rather keep it.

File: f32/f32_amd64.s [hunk 14]
```diff
@@ -4439,3 +4439,277 @@ addsub_scalar:
...
+    VZEROUPPER
+    RET
+
+// func convolveDecimateAVX512(dst, signal, kernel []float32, factor, phase int)
+//
+// AVX-512 fused decimating valid convolution. Inner dot replicates
+// dotProductAVX512 (4 Z accumulators, 64/16/scalar reduction) for bit-identical
+// results vs a per-window DotProductUnsafe on the AVX-512 path.
+TEXT ·convolveDecimateAVX512(SB), NOSPLIT, $0-88
+    MOVQ dst_base+0(FP), R8
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f32/f32_amd64.s:4553]: Keeping R8, same reasoning as the AVX kernel: R8 is a valid scratch register (not reserved like R14), the DX swap is an immeasurable code-size change, and I'm keeping the documented R8-R13 outer-state grouping.

File: f32/f32_amd64.s [hunk 15]
```diff
@@ -4439,3 +4439,277 @@ addsub_scalar:
...
+    VZEROUPPER
+    RET
+
+// func convolveDecimateSSE(dst, signal, kernel []float32, factor, phase int)
+//
+// SSE2 fused decimating valid convolution. Inner dot replicates dotProductSSE
+// (single accumulator, 4-wide loop, SHUFPS horizontal sum) for bit-identical
+// results vs a per-window DotProductUnsafe on the SSE path.
+TEXT ·convolveDecimateSSE(SB), NOSPLIT, $0-88
+    MOVQ dst_base+0(FP), R8
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f32/f32_amd64.s:4654]: Keeping R8, same reasoning as the AVX kernel: R8 is a valid scratch register (not reserved like R14), the DX swap is an immeasurable code-size change, and I'm keeping the documented R8-R13 outer-state grouping.

File: f64/f64_amd64.s [hunk 16]
```diff
@@ -3921,3 +3921,267 @@ round_avx_scalar:
...
+//
+// Fused decimating valid convolution. For each output k it computes the dot
+// product of signal[pos:pos+kLen] with kernel, then advances pos by factor.
+// The inner dot replicates dotProductAVX exactly (4 accumulators, 16/4/scalar
+// reduction) so results are bit-identical to a per-window DotProductUnsafe.
+// Outer state lives in R8-R13 plus BX (pos); the inner dot uses SI/DI/CX/AX and
+// Y0-Y5. BX (not R14) holds pos: R14 is the goroutine g pointer in Go's amd64
+// ABI and must not be clobbered.
+TEXT ·convolveDecimateAVX(SB), NOSPLIT, $0-88
+    MOVQ dst_base+0(FP), R8        // output pointer
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f64/f64_amd64.s:3935]: Keeping R8, same reasoning as the f32 kernels: R8 is a valid scratch register (not reserved like R14), the DX swap is an immeasurable code-size change, and I'm keeping the documented R8-R13 outer-state grouping.

File: f64/f64_amd64.s [hunk 17]
```diff
@@ -3921,3 +3921,267 @@ round_avx_scalar:
...
+    VZEROUPPER
+    RET
+
+// func convolveDecimateAVX512(dst, signal, kernel []float64, factor, phase int)
+//
+// AVX-512 fused decimating valid convolution. Inner dot replicates
+// dotProductAVX512 (4 Z accumulators, 32/8/scalar reduction) for bit-identical
+// results vs a per-window DotProductUnsafe on the AVX-512 path.
+TEXT ·convolveDecimateAVX512(SB), NOSPLIT, $0-88
+    MOVQ dst_base+0(FP), R8
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f64/f64_amd64.s:4033]: Keeping R8, same reasoning as the f32 kernels: R8 is a valid scratch register (not reserved like R14), the DX swap is an immeasurable code-size change, and I'm keeping the documented R8-R13 outer-state grouping.

File: f64/f64_amd64.s [hunk 18]
```diff
@@ -3921,3 +3921,267 @@ round_avx_scalar:
...
+    RET
+
+// func convolveDecimateSSE2(dst, signal, kernel []float64, factor, phase int)
+//
+// SSE2 fused decimating valid convolution (also used on the AVX-without-FMA
+// path, which selects the SSE2 dot). Inner dot replicates dotProductSSE2
+// (single accumulator, 2-wide loop, SHUFPD horizontal sum) for bit-identical
+// results vs a per-window DotProductUnsafe on those paths.
+TEXT ·convolveDecimateSSE2(SB), NOSPLIT, $0-88
+    MOVQ dst_base+0(FP), R8
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f64/f64_amd64.s:4134]: Keeping R8, same reasoning as the f32 kernels: R8 is a valid scratch register (not reserved like R14), the DX swap is an immeasurable code-size change, and I'm keeping the documented R8-R13 outer-state grouping.
